packages/opentelemetry-basic-tracer/src/Span.ts

```diff
@@ -123,17 +124,24 @@ export class Span implements types.Span {
     return true;
   }
 
-  toString() {
-    const json = JSON.stringify({
+  toReadableSpan(): ReadableSpan {
```

Comment:
As discussed, the readable span should be the same object (so basically `this`) but casted to `ReadableSpan`. Otherwise if becomes impossible to access any metadata attached by the vendor.

Reply:
Could we make it so that the `Span` class here also has something like `implements types.WritableSpan, types. ReadableSpan` so that we ensure it has both the read and write methods?

Reply:
> Otherwise it becomes impossible to access any metadata attached by the vendor.

How do you plan to attach metadata? One way I can think is adding new properties on `Span` class. Could you please elaborate more?

> Could we also make it so that the Span class here also has something like implements types.WritableSpan, types. ReadableSpan

I am not sure if this is possible, Properties are private in type `Span`, but not in type `ReadableSpan`.

Reply:
Well, here is how we would do it:
- Make properties public on `Span` here
- Make the actual exported methods on `Tracer` like `startSpan` return a TypeScript interface type of `WriteableSpan` showing that we do not intend for users to access any of the span methods, and we document them as such.
- Make the type of `Span` passed to the exporter be `ReadableSpan`, which only has the read properties accessible.
- Don't export the `Span` type from the `basic-tracer` package, so end users can't access the class directly but only through the associated interfaces.

Of course people writing in JS (or TS folks who cast the type) can of course access the properties - but that is true of private properties anyway. What we are doing with the types and the comments on the properties is explaining the *intended* use of the API, such that a user who deviates from that will know they are doing something non-supported.

What do you think of that approach?

Reply:
I think it makes sense. My only worry with `WritableSpan` is that it gives the impression that it cannot be read from at all, which may not be true. But as far as the idea itself, I think it solves the problem completely.

> Could you please elaborate more?

What I meant is that it needs to be the same underlying object, regardless of the interface. Then it allows the vendor properties to be retrieved and altered as needed. While it would be possible to assign the vendor specific data to multiple different objects, it's possible one of the things needed from the vendor is also the span itself, in which case we would end up with a circular dependency. It's best if we can limit what is available through types, but keep a single object that is passed around.

Reply:
Just to reiterate a point mentioned during the SIG meeting, the API interface for [types.Span](https://github.com/open-telemetry/opentelemetry-js/blob/master/packages/opentelemetry-types/src/trace/span.ts#L28) is already a write only span and is what's returned by startSpan. That being said if you aren't using ts, just pure js, the user isn't forced to follow the interface typing../

Reply:
@draffensperger @rochdev @bg451 thanks for the reviews and suggestions. I think I have addressed all the comments. PTAL.

Reply:
Regrading JS users being able to override the values directly - that's true, but if they are following the docs/reading the APIs they would likely recognize that that isn't the intended approach. But in JS you are really free - you can read private properties, monkey patch, etc. (and in TS the same is true as well, you just need some casting). So I think we balance prevention of issues, performance and clean API - and this design to me at least for now feels like a reasonable trade-off of those different things.